Index: binutils-2.16.1/binutils/dlltool.c
===================================================================
--- binutils-2.16.1.orig/binutils/dlltool.c
+++ binutils-2.16.1/binutils/dlltool.c
@@ -389,7 +389,7 @@ static int verbose;
 static FILE *output_def;
 static FILE *base_file;
 
-#ifdef DLLTOOL_ARM
+#if defined(DLLTOOL_ARM) && !defined(DLLTOOL_I386)
 #ifdef DLLTOOL_ARM_EPOC
 static const char *mname = "arm-epoc";
 #else
