---
 c/src/lib/libbsp/m68k/acinclude.m4 |    2 ++
 1 file changed, 2 insertions(+)

Index: rtems/c/src/lib/libbsp/m68k/acinclude.m4
===================================================================
--- rtems.orig/c/src/lib/libbsp/m68k/acinclude.m4	2007-08-14 18:59:45.000000000 +0200
+++ rtems/c/src/lib/libbsp/m68k/acinclude.m4	2007-08-14 19:04:40.000000000 +0200
@@ -20,6 +20,8 @@
     AC_CONFIG_SUBDIRS([mcf5206elite]);;
   mcf5235 )
     AC_CONFIG_SUBDIRS([mcf5235]);;
+  mo376 )
+    AC_CONFIG_SUBDIRS([mo376]);;
   mrm332 )
     AC_CONFIG_SUBDIRS([mrm332]);;
   mvme136 )
