---
 c/src/lib/libbsp/m68k/acinclude.m4 |    2 ++
 1 file changed, 2 insertions(+)

Index: rtems-4.9.3/c/src/lib/libbsp/m68k/acinclude.m4
===================================================================
--- rtems-4.9.3.orig/c/src/lib/libbsp/m68k/acinclude.m4	2008-08-30 22:29:11.000000000 +0200
+++ rtems-4.9.3/c/src/lib/libbsp/m68k/acinclude.m4	2009-11-29 01:35:52.795572383 +0100
@@ -24,6 +24,8 @@
     AC_CONFIG_SUBDIRS([mcf5235]);;
   mcf5329 )
     AC_CONFIG_SUBDIRS([mcf5329]);;
+  mo376 )
+    AC_CONFIG_SUBDIRS([mo376]);;
   mrm332 )
     AC_CONFIG_SUBDIRS([mrm332]);;
   mvme136 )
