---
 cpukit/libcsupport/src/readlink.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: rtems/cpukit/libcsupport/src/readlink.c
===================================================================
--- rtems.orig/cpukit/libcsupport/src/readlink.c	2007-08-15 01:54:35.000000000 +0200
+++ rtems/cpukit/libcsupport/src/readlink.c	2007-08-15 01:56:45.000000000 +0200
@@ -18,10 +18,10 @@
 #include <rtems/libio_.h>
 #include <rtems/seterr.h>
 
-int readlink(
+ssize_t readlink(
   const char *pathname,
   char       *buf,
-  int         bufsize
+  size_t      bufsize
 )
 {
   rtems_filesystem_location_info_t  loc;
