Index: rtems/cpukit/libcsupport/include/inttypes.h
===================================================================
--- rtems.orig/cpukit/libcsupport/include/inttypes.h
+++ rtems/cpukit/libcsupport/include/inttypes.h
@@ -281,8 +281,8 @@ extern intmax_t  imaxabs(intmax_t j);
 extern imaxdiv_t imaxdiv(intmax_t numer, intmax_t denomer);
 extern intmax_t  strtoimax(const char *__restrict, char **__restrict, int);
 extern uintmax_t strtoumax(const char *__restrict, char **__restrict, int);
-extern intmax_t  wcstoimax(const wchar_t *__restrict, wchar_t **__restrict, int);
-extern uintmax_t wcstoumax(const wchar_t *__restrict, wchar_t **__restrict, int);
+//extern intmax_t  wcstoimax(const wchar_t *__restrict, wchar_t **__restrict, int);
+//extern uintmax_t wcstoumax(const wchar_t *__restrict, wchar_t **__restrict, int);
 
 #ifdef __cplusplus
 }
