Index: rtems-051009/cpukit/libfs/src/dosfs/msdos_misc.c
===================================================================
--- rtems.orig/cpukit/libfs/src/dosfs/msdos_misc.c
+++ rtems/cpukit/libfs/src/dosfs/msdos_misc.c
@@ -252,7 +252,7 @@ msdos_find_name(
  *     error occured (errno set apropriately)
  *
  */
-int
+msdos_status_t
 msdos_get_name_node(
     rtems_filesystem_location_info_t *parent_loc,
     char                             *name,
@@ -762,7 +762,7 @@ msdos_dir_is_empty(
  *     appropriately)
  *
  */
-int
+msdos_status_t
 msdos_find_name_in_fat_file(
     rtems_filesystem_mount_table_entry_t *mt_entry,
     fat_file_fd_t                        *fat_fd,
@@ -880,7 +880,7 @@ msdos_find_name_in_fat_file(
  *     RC_OK on success, or error code if error occured
  *
  */
-int
+msdos_status_t
 msdos_find_node_by_cluster_num_in_fat_file(
     rtems_filesystem_mount_table_entry_t *mt_entry,
     fat_file_fd_t                        *fat_fd,
