Index: rtems-051009/c/src/lib/libbsp/m68k/acinclude.m4
===================================================================
--- rtems.orig/c/src/lib/libbsp/m68k/acinclude.m4
+++ rtems/c/src/lib/libbsp/m68k/acinclude.m4
@@ -22,6 +22,8 @@ AC_DEFUN([RTEMS_CHECK_BSPDIR],
     AC_CONFIG_SUBDIRS([mcf5235]);;
   mrm332 )
     AC_CONFIG_SUBDIRS([mrm332]);;
+  mo376 )
+    AC_CONFIG_SUBDIRS([mo376]);;
   mvme136 )
     AC_CONFIG_SUBDIRS([mvme136]);;
   mvme147 )
